### PR TITLE
Feature/testing for merge into the Dev Branch. 

### PR DESCRIPTION
**Pull request by stravid87**
## Description

INSERT YOUR DESCRIPTION HERE.

---

Before requesting a review of this Pull Request, I certify that I have tested the following items in the application and all have passed my local tests without any side effects:

- [X] Logout and login;
- [X] Navigation between pages;
- [X] Creation of a new prompt;
- [X] Creation of a new entry;
- [X] Like and Dislike button;
- [X] Share button;
- [X] Exclusion of the newly created prompt;
- [X] Deletion of the newly created entry;

---

The purpose here is to allow feature/unit-test to be deleted -- it is now outdated. 

The new branch features/testing is built off of a recent pull from the dev branch and so is more up to date.

### CODE REVIEW

**Comment by stravid87** [comment]
@arnonrdp
I left general instructions at the bottom of Issue #42 as an alternative to directly merging.

**Comment by arnonrdp** [comment]
This is my console after running `npm run test`:

Captura de Tela 2023-02-07 às 18 21 07

**Comment by arnonrdp** [comment]
I noticed you talked about `/firebase.rules` but in this branch we have `/firestore.rules`.
Which one should we use?

**Comment by stravid87** [comment]
Try running `npm run emulators` first. I think it needs the emulators up & running otherwise it gives that msg.

**Comment by stravid87** [comment]
> I noticed you talked about `/firebase.rules` but in this branch we have `/firestore.rules`. Which one should we use?
> 
> By the way, smart decision to mock an Admin so it can create, update and delete prompts and entries.

`/firestore.rules` -- typo
Where did you see this?

**Comment by arnonrdp** [comment]
> Try running `npm run emulators` first. I think it needs the emulators up & running otherwise it gives that msg. Also, change `firebase.js` to not use analytics (you'll see the boolean) but yes use the emulators (also a boolean). Thinking about this now, there is almost certainly a programmatic way to do this....

I got your point. There is a way, no worries.

---

> `/firestore.rules` -- typo Where did you see this?

I saw here: https://github.com/globe-and-citizen/Celebrity-Fanalyzer/issues/42#issuecomment-1416106334

**Comment by arnonrdp** [comment]
I'm running `npm run emulators` first and `npm run test` next.
Here is what I have:

Run Emulators

Run Test

The Firebase Emulator Suite is working pretty well.
I noticed we are not using Hosting, so I think it could be removed (I just don't know the impact and difficult of removing it).

About the above error, doesn't it happen to you?

**Comment by stravid87** [comment]
Huh... no. I just pulled the latest  with your changes and it works.

**Comment by stravid87** [comment]
@arnonrdp When you run the emulators, are you pulling in all the other fake data? Are you able to: 1) run emulators, 2) run `npm run dev` and then add and interact with the emulators?

**Comment by stravid87** [comment]
So I'm able to succeed the test and fail the test by changing the user email in the string
```
[L21 /MainMenu.spec.js]
const userString = '{"sub": "whatSubout?", "email": "algae.peach.153@example.com", "email_verified": true}'
```
unfortunately, I don't know how to sign out programmatically. I need to shut down the emulator and restart otherwise, a previously signed in user persists despite changing the string above

**Comment by arnonrdp** [comment]
Yes, I am able to interact with fake data:

Fake Prompts

The error I have here is: **FirebaseError: Firebase: Error (auth/network-request-failed)**. In the [docs](https://firebase.google.com/docs/reference/js/v8/firebase.auth.Error) it says:
> **auth/network-request-failed**
> Thrown if a network error (such as timeout, interrupted connection or unreachable host) has occurred.

I believe this has something to do with the connection ports.

---

To logout programmatically I believe we should just call `authStore.logout()` at the end of everything.

**Comment by arnonrdp** [comment]
The error occurs when calls this function: `await signInWithCredential(auth, credential)`.

I added these 3 console.log, so you can see it returns credential (but the token is _null_) but it doesn't return the result.

Logs

I gave I password to the user and tried to use `signInWithEmailAndPassword(auth, email, password)` but the error returned is the same.


@hermannleboss, could you please test this? If you don't face any issues, let's go ahead because my intention is to implement these tests in a GitHub Action.

**Comment by stravid87** [comment]
ahhhh. when you run emulators, what ports are used? 

```
│ Authentication │ 127.0.0.1:9099 │ http://127.0.0.1:4001/auth      │
├────────────────┼────────────────┼─────────────────────────────────┤
│ Firestore      │ 127.0.0.1:8000 │ http://127.0.0.1:4001/firestore │
├────────────────┼────────────────┼─────────────────────────────────┤
│ Hosting        │ 127.0.0.1:5000 │ n/a                             │
├────────────────┼────────────────┼─────────────────────────────────┤
│ Storage        │ 127.0.0.1:9199 │ http://127.0.0.1:4001/storage   │
```

**Comment by stravid87** [comment]
Making progress. I'm going to see if I can write tests to add / delete prompts & entries tomorrow. 

```
describe('Main Menu Component', async () => {
  beforeEach(async () => {
    setActivePinia(createPinia())
    localStorage.clear()
  })

  it('is should NOT display the admin panel if the user logged has the admin role', async () => {
    const userStore = useUserStore()
    const credential = await createUserWithEmailAndPassword(auth, `user_${Math.round(Math.random() * 100)}@gstring.com`, '12341234')
    const isNewUser = getAdditionalUserInfo(credential)?.isNewUser
    const { email, displayName, photoURL, uid } = credential.user
    console.log('USER: ', email, displayName, photoURL, uid)
    if (isNewUser) await setDoc(doc(db, 'users', uid), { email, displayName, photoURL })
    await userStore.fetchUserProfile(credential.user)

    const admin = userStore.isAdmin
    expect(admin).toBeFalsy()

    let localUserString = localStorage.getItem('user')
    expect(localUserString.indexOf('role')).toBe(-1)

    const wrapper = mount(MainMenu)
    expect(wrapper.find('[to="/admin"]').exists()).toEqual(false)
  })

  it('it SHOULD display the admin panel if the user logged has the admin role', async () => {
    const userStore = useUserStore()
    const userString = '{"sub": "what_is_sub?", "email": "algae.peach.153@example.com", "email_verified": true}'
    const credential = GoogleAuthProvider.credential(userString)
    console.log('Credential: ', credential)
    const result = await signInWithCredential(auth, credential)
    const { email, displayName, photoURL, uid } = result.user
    console.log('Should be algae.peach? Got: ', email, displayName, photoURL, uid)
    /*set in localstroage */ await userStore.fetchUserProfile(result.user)
    const admin = userStore.isAdmin
    console.log('admin', admin)
    let localUserString = localStorage.getItem('user')
    expect(localUserString.indexOf('role')).toBeTruthy()

    const wrapper = mount(MainMenu)
    expect(wrapper.find('[to="/admin"]').exists()).toEqual(true)
  })
```

**Comment by stravid87** [comment]
> sterday I needed to make some improvements related to authentication (now it is too

After doing this, did you get the error to resolve? Where you able to work with the auth emulator?

File: CONTRIBUTING.md [hunk 1]
```diff
@@ -37,6 +37,17 @@ It is not necessary to configure anything related to this.
...
 ```
 
+### Test your code
+
+Before submitting your PR you need to make sure that nothing has been broken.
+For that, you need to run the tests:
+
+```bash
+npm install -g firebase-tools # In case you don't have firebase-tools installed
+npm run emulators
```
Comment (hermannleboss) [inline CONTRIBUTING.md:47]: Before add 
`firebase login # to authenticate the CLI`

**Comment by stravid87** [comment]
changing the .env file VITE_RELEASE_STAGE=development to testing resolved the problem. 
`FirebaseError: Firebase: Unable to parse Google id_token: {"sub": "what_is_sub?", "email": "algae.peach.153@example.com", "email_verified": true} (auth/invalid-credential).`

**Comment by stravid87** [comment]
I'm now running into some challenges in Fort McMurray running the tests with emulators -- this will be tomorrow's work.

**Comment by arnonrdp** [comment]
I just solved the conflicts.
Do you mind adding any queries to the database that don't require authentication?
It could be get or like prompts.

**Comment by stravid87** [comment]
Closed in favour of the testing_for_merge pull request.

**Pull request closed without merging**